packages/eslint-plugin/src/rules/no-restricted-imports.ts

```diff
@@ -0,0 +1,185 @@
+            paths: schemaForMergeArrayOfStringsOrObjects,
+            patterns: schemaForMergeArrayOfStringsOrObjectPatterns,
+          },
+        },
+      },
+    ],
+  },
+);
+
+function isObjectOfPaths(
+  obj: unknown,
+): obj is { paths: ArrayOfStringOrObject } {
+  return Object.prototype.hasOwnProperty.call(obj, 'paths');
+}
+
+function isObjectOfPatterns(
+  obj: unknown,
+): obj is { patterns: ArrayOfStringOrObjectPatterns } {
+  return Object.prototype.hasOwnProperty.call(obj, 'patterns');
+}
+
+function isOptionsArrayOfStringOrObject(
+  options: Options,
+): options is ArrayOfStringOrObject {
+  if (isObjectOfPaths(options[0])) {
+    return false;
+  }
+  if (isObjectOfPatterns(options[0])) {
+    return false;
+  }
+  return true;
+}
+
+function getRestrictedPaths(options: Options): ArrayOfStringOrObject {
+  if (isOptionsArrayOfStringOrObject(options)) {
+    return options;
+  }
+  if (isObjectOfPaths(options[0])) {
+    return options[0].paths;
+  }
+  return [];
+}
+
+function getRestrictedPatterns(
+  options: Options,
+): ArrayOfStringOrObjectPatterns {
+  if (isObjectOfPatterns(options[0])) {
+    return options[0].patterns;
+  }
+  return [];
+}
+
+export default createRule<Options, MessageIds>({
+  name: 'no-restricted-imports',
+  meta: {
+    type: 'suggestion',
+    docs: {
+      description: 'Disallow specified modules when loaded by `import`',
+      category: 'Best Practices',
+      recommended: false,
+      extendsBaseRule: true,
+    },
+    messages: baseRule.meta.messages,
+    fixable: baseRule.meta.fixable,
+    schema,
+  },
+  defaultOptions: [],
+  create(context) {
+    const rules = baseRule.create(context);
+    const { options } = context;
+
+    const restrictedPaths = getRestrictedPaths(options);
+    const allowedTypeImportPathNameSet: Set<string> = new Set();
+    for (const restrictedPath of restrictedPaths) {
+      if (
+        typeof restrictedPath === 'object' &&
+        restrictedPath.allowTypeImports
+      ) {
+        allowedTypeImportPathNameSet.add(restrictedPath.name);
+      }
+    }
+    function isAllowedTypeImportPath(importSource: string): boolean {
+      return allowedTypeImportPathNameSet.has(importSource);
+    }
+
+    const restrictedPatterns = getRestrictedPatterns(options);
+    const allowedImportTypeMatchers: Ignore[] = [];
+    for (const restrictedPattern of restrictedPatterns) {
+      if (
+        typeof restrictedPattern === 'object' &&
+        restrictedPattern.allowTypeImports
+      ) {
+        allowedImportTypeMatchers.push(ignore().add(restrictedPattern.group));
+      }
+    }
+    function isAllowedTypeImportPattern(importSource: string): boolean {
+      return allowedImportTypeMatchers.every(matcher => {
+        return matcher.ignores(importSource);
+      });
+    }
+
+    return {
+      ImportDeclaration(node): void {
+        if (node.importKind === 'type') {
+          const importSource = (node.source.value as string).trim();
```

Comment:
nit - right now our parser allows non-string `source.value`'s - so we need to handle this gracefully.

ideally we can just quit early if the value isn't a string.